feat: Implement _initiateRelayerRefund and begin Optimism-specific SpokePool

nicholaspai opened the pull request.

chrismaree commented on contracts/SpokePool.sol:
```diff
@@ -43,6 +43,16 @@ abstract contract SpokePool is Testable, Lockable, MultiCaller {
+    uint256 public relayerRefundRootCount;
+
+    // This is a 2D bitmap tracking which leafs in the relayer refund roots have been claimed, with max size of
+    // 256 elements, limiting us to 256 relayer refund leafs per distribution proof.
```
this comment is wrong. we have no such limitation.

nicholaspai replied:
fixed

chrismaree commented on contracts/SpokePool.sol:
```diff
@@ -257,7 +267,15 @@ abstract contract SpokePool is Testable, Lockable, MultiCaller {
+    // specifics are left to the implementor of this abstract contract.
+    function _initializeRelayerRefund(bytes32 relayerRepaymentDistributionProof) internal {
+        relayerRefunds.push();
+        uint256 relayerRefundId = relayerRefunds.length - 1;
```
small gas optitimization: get the `uint256 relayerRefundId = relayerRefunds.length;` before doing the push operation so you dont need to do the extra subtraction operation.

nicholaspai commented:
@mrice32 @chrismaree not sure why but I the `yarn.lock` wasn't regenerating after I added something to the package.json, so i removed and then re-added and it caused a big diff

mrice32 commented on yarn.lock:
```diff
@@ -279,20 +279,20 @@
-  version "0.5.8"
-  resolved "https://registry.yarnpkg.com/@eth-optimism/contracts/-/contracts-0.5.8.tgz#975313a219b5537bad7f4f98f8841c98592882d0"
-  integrity sha512-INvIMiE72Y6tjxm+BZyN9f3+LiMKJgmNpR0qEimrvpg49hK8gOXQTnI+0C7/qIIGvgaQgDyejG9AfkxZizjLVQ==
+"@eth-optimism/contracts@^0.5.11", "@eth-optimism/contracts@^0.5.5":
```
@nicholaspai the reason you didn't see changes in the yarn.lock (I think) was that this package was _already in the dependency tree_ with a permissible version, most likely due to the fact that we import uma/core. You could probably drop these yarn.lock changes if you wanted and I think everything would work (but nothing wrong with regenerating things at this early stage).

nicholaspai replied:
interesting, what do you think we should do--keep the new yarn.lock or drop? Also should we update the eth-optimism version if its available?

mrice32 replied:
I think keeping the updates is totally fine! I just wanted to figure out what happened to make sure something wasn't broken in our setup or something :)

mrice32 commented on contracts/Optimism_SpokePool.sol:
```diff
@@ -0,0 +1,65 @@
+     *          ADMIN FUNCTIONS           *
+     **************************************/
+
+    // TODO:
```
What's TODO on these? The natspec or something in the implementation?

nicholaspai replied:
yeah i guess these are fully implemented actually

mrice32 commented on contracts/Optimism_SpokePool.sol:
```diff
@@ -0,0 +1,65 @@
+     *         INTERNAL FUNCTIONS         *
+     **************************************/
+
+    function _setCrossDomainAdmin(address newCrossDomainAdmin) internal {
```
Is there any way to call this function after construction? Is this something we'd like to be able to change? (genuine question, not suggesting a change)

chrismaree replied:
there is a `setCrossDomainAdmin` that can call this after construction enabling us to change it. I think we'd want to change it in the event something broke within the L1<->L2 logic. one problem, though, that I can see with this pattern is it assumes that the cross domain owner implements sufficient methods to call `setCrossDomainAdmin` over the bridge. this should be doable as this is going to be sitting within the "Adapter" logic that I've created for the L1->L2 calls. We can create a base contract (`CrossDomainAdminBase`) or something equivalent that these adapters will inherit from. can think more on this in the next few PRs.

nicholaspai replied:
Yeah i added in this function without telling you. We should allow this to be changed from L1

mrice32 replied:
Sorry, I'm a little confused. Are there two functions that allow you to `setCrossDomainAdmin`? If so, can we consolidate them into a single function?

nicholaspai replied:
No there's only one, implemented in this contract. `setCrossDomainAdmin` is included in the `SpokePoolInterface`

mrice32 commented on contracts/Optimism_SpokePool.sol:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    // TODO:
+    function setDepositQuoteTimeBuffer(uint64 buffer) public onlyFromCrossDomainAccount(crossDomainAdmin) {
```
Is there any interface or anything that ensures that these functions match across all SpokePool implementations?

nicholaspai replied:
Great point I'll add an interface

nicholaspai replied:
do you think all L2 spoke pools will immplement the notion of a `crossDomainAdmin`? @chrismaree

mrice32 reviewed (approved):
Looks good! Only minor questions/comments

chrismaree commented on test/constants.ts:
```diff
@@ -39,3 +39,7 @@ export const zeroBytes32 = "0x00000000000000000000000000000000000000000000000000
 
 export const zeroRawValue = { rawValue: "0" };
+
+export const spokePoolRelayerRefundRoot = createRandomBytes32();
```
I have a const for this in my other PR called `mockDestinationDistributionRoot`.

nicholaspai replied:
Will merge that in

nicholaspai commented on contracts/Optimism_SpokePool.sol:
```diff
@@ -0,0 +1,80 @@
+
+contract Optimism_SpokePool is CrossDomainEnabled, SpokePoolInterface, SpokePool {
+    // Address of the L1 contract that acts as the owner of this SpokePool.
+    address public override crossDomainAdmin;
```
@chrismaree @mrice32 should `crossDomainAdmin` be in the `SpokePool` or do you think this notion of a cross domain admin will be different for each. L2?

nicholaspai replied:
Its the contract that calls `relayMessage` on the canonical bridge, so its either the `BridgeAdmin` or the `HubPool`

chrismaree replied:
for now `HubPool` me thinks. we might want to change this later but I think lets leave it there optimistically at the moment.

mrice32 reviewed (approved):
LGTM!

nicholaspai commented on contracts/SpokePool.sol:
```diff
@@ -43,6 +45,15 @@ abstract contract SpokePool is Testable, Lockable, MultiCaller {
 
+    struct RelayerRefund {
+        // Merkle root of relayer refunds.
+        bytes32 distributionRoot;
```
Hm so you think this could work by the `HubPool` relaying all leafs to the `SpokePool` here and storing them on-chain, as opposed to just relaying the root. @chrismaree wdyt?

chrismaree replied:
I dont think @mrice32  is proposing the HubPool passing the leafs; it cant do that in a efficient way. I _think_ @mrice32 is proposing a technique where you can execute multiple leafs at once in a compressed structure and prove they are within the provided root. @mrice32 what you describe would fit into the execution stage of the process, right?

nicholaspai replied:
How do you imagine we improve on just making the contract multicall and assuming `distributeRelayerRefund` is called multiple times in one batched txn?

mrice32 replied:
I wanna be super clear that I don't think we should do this now, and I think before we add this complexity, we should try to better understand the amount of calldata/gas savings we'd see from something like this with different tree sizes.

Agree @chrismaree. I'm not suggesting the HubPool do anything different since any added data anywhere on mainnet is going to be expensive.

If you did multicall, you would effectively be passing:

[{distributionStruct1, proofArray1}, {distributionStruct2, proofArray2}...]

What I'm suggesting is an additional method that allows you to pass _all_ leaves this way:

[{distriubtionStruct1}, {distributionStruct2}, ...]

If we get the logic right in the contract, the leaf data _could be used_ to generate the proofs so you wouldn't need to pass them in, thereby reducing calldata.

chrismaree commented on test/SpokePool.RelayerRefund.ts:
```diff
@@ -0,0 +1,22 @@
+import { expect } from "chai";
```
In your next PR it might be good to re-name this file to `SpokePool.RefundInitialization.ts` and then you can create `SpokePool.RefundExecution.ts` for the execution step, following the pattern I did in the hub.

nicholaspai replied:
done!

chrismaree reviewed (approved):
this looks great to me! keen to get the next bit of functionality in with the proofs.

nicholaspai commented:
Just giving @mrice32 a chance to comment [here](https://github.com/across-protocol/across-smart-contracts-v2/pull/20#discussion_r800108266) before merging
